File: Week 1/id_005/LeetCode-141-005.go
/**
 * 使用快慢指针
 * 执行用时 :8 ms, 在所有 golang 提交中击败了90.53%的用户
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func hasCycle(head *ListNode) bool {
	if head == nil {
		return false
	}
	// 设置快慢指针
	slow, fast := head, head
	for fast.Next != nil && fast.Next.Next != nil {
		fast = fast.Next.Next
		slow = slow.Next
		if slow == fast {
			return true
		}
	}
	return false
}

func hasCycleOne(head *ListNode) bool {
    if head == nil || head.Next == nil {
        return false
    }
    fast, slow := head.Next, head
    for fast != slow {
        if fast == nil || fast.Next == nil {
            return false
        }
        fast = fast.Next.Next
        slow = slow.Next
    }

    return true
}

/**
 * 迭代
 * 执行用时 :8 ms, 在所有 golang 提交中击败了90.05%的用户
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func hasCycleTwo(head *ListNode) bool {
    for head != nil {
        if head == head.Next {
            return true
        }
        if head.Next != nil {
            head.Next = head.Next.Next
        }
        head = head.Next
    }
    return false
}

File: Week 1/id_005/LeetCode-141-005.php
<?php
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
class Solution
{
    /**
     * 快慢指针
     * @param ListNode $head
     * @return boolean
     */
    public function hasCycle($head)
    {
        if ($head == null || $head->next == null) {
            return false;
        }
        $slow = $head;
        $fast = $head->next;
        while ($fast != $slow) {
            if ($fast == null || $fast->next == null) {
                return false;
            }
            $slow = $slow->next;
            $fast = $fast->next->next;
        }
        return true;
    }

    /**
     * 迭代
     * @param ListNode $head
     * @return boolean
     */
    public function hasCycle($head)
    {
        while ($head != null) {
            if ($head == $head->next) {
                return true;
            }
            if ($head->next != null) {
                $head->next = $head->next->next;
            }
            $head = $head->next;
        }
        return false;
    }
}

class ListNode
{
    public $val;
    public $next;
    public function __construct($val = 0, $next = null)
    {
        $this->val  = $val;
        $this->next = $next;
    }
}

class List
{
    public $val;
    public $next;

    public function __construct($arrs, $p = -1)
    {
        $this->val = $arrs[0];
        $pre = $this;
        for ($i = 1; $i < count($arrs); $i++) {
            $pre->next = new ListNode();
            $pre->next->val = $arrs[$i];
            $pre = $pre->next;
            if ($p == 1 && $i == count($arrs) - 1) {
                $pre->next = $this->next;
            }
            if ($p == 0 && $i == count($arrs) - 1) {
                $pre->next = $this;
            }
        }
    }
}
